docs/concepts/style_guide.md

```diff
@@ -0,0 +1,216 @@
+```
+- Any exit code allowed
+```ruby
+on(host, "rm #{file_to_rm}", :accept_all_exit_codes => true)
+```
+
+In the last case, when any exit code is allowed, one must follow-up with a valid
+assertion test.
+
+If an exit_code outside of 0 is expected, one must use acceptable_exit_codes so
+the test will fail on the proper assertion and not error at that command.  Allow
+only the minimum expected set of exit codes unless coverage is provided by
+subsequent assertions.
+
+## Test Outcomes
+
+When to use each, and how to format the message:
+
+### Expecting Failure
+
+If your tests are failing due to an "expected failure", you should wrap your
+failing assertion in an
+[`expect_failure`](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Structure#expect_failure-instance_method)
+block with an explanatory logging message:
+
+```ruby
+expect_failure('expected to fail due to PE-1234') do
+  assert_equal(400, response.code, 'bad response code from API call')
+end
+```
+
+Note that `expect_failure` will only trigger from failed assertions. It won't
+take care of failed host or `on` commands. To deal with expected failure from an
+`on` invocation, you'd want something more like this:
+
+```ruby
+on(blah, 'blah', :allow_all_exit_codes => true) do |result|
+  expect_failure 'known issue: TIK-1234' do
+    assert_equal(4,result.exit_code,'did not receive expected exit code for blah')
+  end
+end
+```
+
+### `fail_test`, `pass_test`, & `skip_test`
+
+These can be used anywhere in a test to exit early. A
+[`skip_test`](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Outcomes#fail_test-instance_method)
+between two assertions, for instance, will run the first assertion, raise an
+exception for `skip_test`, run teardown, and then exit as we expect. The same is
+true for
+[fail_test](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Outcomes#fail_test-instance_method).
+
+[`pass_test`](http://www.rubydoc.info/github/puppetlabs/beaker/Beaker/DSL/Outcomes#fail_test-instance_method)
+is typically not required. When the end of a test is reached without causing an
+error (due to bad test code execution, or an unhandled exit code or exception)
+or failure (due to assertions), then it passes.
+
+`pass_test` can be used in a situation where one knows a test has passed before
+the end of a test under certain circumstances, such as during a loop that has
+not yet completed.
+
+### Skipping Tests
+
+Skipping tests can be used, for instance, when they are temporarily failing or
+not yet complete.
+
+**Good:**
+```ruby
+skip_test 'requires puppet and mcollective service scripts from AIO agent package' if @options[:type] != 'aio'
+```
+**Bad:**
+```ruby
+confine :to, :platform => 'solaris:pending'
+```
+
+## Assertions
+
+Always include a _unique_ error message in your assertion statement. Use strict
+asserts whenever possible (e.g. assert_equal, assert_match.
+[More info](http://danwin.com/2013/03/ruby-minitest-cheat-sheet/)).
+
+## Confining
```

Comment:
It seems to me that this confine section should just be moved to the `Skipping Tests` section, as there is code that is using `confine` there, but not real explanation given on why it is "bad".